tscompare/methods.py

```diff
@@ -0,0 +1,405 @@
+
+    nil = frozenset()
+
+    # Initialize clade iterators
+    query = CladeMap(ts)
+    target = CladeMap(other)
+
+    # Initialize buffer[clade] = (query_nodes, target_nodes, left_coord)
+    modified = query.clades() | target.clades()
+    buffer = {}
+
+    # Build sparse matrix of matches in triplet format
+    query_node = []
+    target_node = []
+    shared_span = []
+    right = 0
+    while True:
+        left = right
+        right = min(query.interval[1], target.interval[1])
+
+        # Flush pairs of nodes that no longer have matching clades
+        for clade in modified:  # flush:
+            if clade in buffer:
+                n_i, n_j, start = buffer.pop(clade)
+                span = left - start
+                for i, j in product(n_i, n_j):
+                    query_node.append(i)
+                    target_node.append(j)
+                    shared_span.append(span)
+
+        # Add new pairs of nodes with matching clades
+        for clade in modified:
+            assert clade not in buffer
+            if clade in query and clade in target:
+                n_i, n_j = query[clade], target[clade]
+                buffer[clade] = (n_i, n_j, left)
+
+        if right == ts.sequence_length:
+            break
+
+        # Find difference in clades with advance to next tree
+        modified.clear()
+        for clade_map in (query, target):
+            if clade_map.interval[1] == right:
+                clade_diff = clade_map.next()
+                for prev, curr in clade_diff.values():
+                    if prev != nil:
+                        modified.add(prev)
+                    if curr != nil:
+                        modified.add(curr)
+
+    # Flush final tree
+    for clade in buffer:
+        n_i, n_j, start = buffer[clade]
+        span = right - start
+        for i, j in product(n_i, n_j):
+            query_node.append(i)
+            target_node.append(j)
+            shared_span.append(span)
+
+    numer = scipy.sparse.coo_matrix(
+        (shared_span, (query_node, target_node)),
+        shape=(ts.num_nodes, other.num_nodes),
+    ).tocsr()
+
+    return numer
+
+
+def match_node_ages(ts, other):
+    """
+    For each node in `ts`, return the age of a matched node from `other`.  Node
+    matching is accomplished by calculating the intervals over which pairs of
+    nodes (one from `ts`, one from `other`) subtend the same set of samples.
+
+    Returns three vectors of length `ts.num_nodes`: the age of the best
+    matching node in `other` (e.g.  with the longest shared span); the
+    proportion of the node span in `ts` that is covered by the best match; and
+    the id of the best match in `other`.
+
+    If either tree sequence contains unary nodes, then there may be multiple
+    matches with the same span for a single node. In this case, the returned
+    match is the node with the smallest integer id.
+    """
+
+    shared_spans = shared_node_spans(ts, other)

```

Comment:
TODO: refactor `compare` to use this (but doing tie-resolving using times as in that function).